### PR TITLE
FIX-#5737: BUG: String columns are converted to Categorical, if exported from HDK

### LINKED ISSUE #5737: BUG: String columns are converted to Categorical, if exported from HDK
### Modin version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the latest released version of Modin.

- [X] I have confirmed this bug exists on the main branch of Modin. (In order to do this you can follow [this guide](https://modin.readthedocs.io/en/stable/getting_started/installation.html#installing-from-the-github-master-branch).)

### Reproducible Example

```python
import modin.pandas as pd

df = pd.DataFrame({"A": ["A"]})
dtypes = df.dtypes
df = df[df["A"] != ""]
assert all(dtypes == df._to_pandas().dtypes)
```

### Issue Description

If an arrow table is exported from HDK, the string columns are converted to dictionary. On conversion to pandas, these columns will be of type Categorical, that is not correct.

### Expected Behavior

The columns types must not be changed to Categorical.

### Error Logs

<details>

```python-traceback

Replace this line with the error backtrace (if applicable).

```

</details>

### Installed Versions

<details>

Replace this line with the output of pd.show_versions()

</details>

### PR DESCRIPTION
**Pull request by AndreyPavlenko**
- [x] first commit message and PR title follow format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
  > **_NOTE:_**  If you edit the PR title to match this format, you need to add another commit (even if it's empty) or amend your last commit for the CI job that checks the PR title to pick up the new PR title.
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #5737 
- [x] tests added and passing
- [ ] module layout described at `docs/development/architecture.rst` is up-to-date

### CODE REVIEW

File: modin/experimental/sql/hdk/query.py [hunk 1]
```diff
@@ -66,7 +67,14 @@ def hdk_query(query: str, **kwargs) -> pd.DataFrame:
     worker = HdkWorker()
     if len(kwargs) > 0:
         query = _build_query(query, kwargs, worker.import_arrow_table)
-    return from_arrow(worker.executeDML(query))
+    at = worker.executeDML(query)
+    schema = at.schema
+    # HDK returns strings as dictionary. Cast dictionary columns to string.
+    if cast := {i: f for i, f in enumerate(schema) if pa.types.is_dictionary(f.type)}:
```
Comment (dchigarev) [inline modin/experimental/sql/hdk/query.py:73]: so we unconditionally cast all dictionary columns to strings, what if there were columns that have to be a dictionary encoded?
```python
mdf = mdf.astype("category")
res = execute_some_query(mdf)
res.dtypes == "string" # ?
```

could please you also measure what it cost us to cast the columns to strings? it feels to be very expensive. Can we implement something like a delayed casting? 

As far as I understand there's nothing wrong for strings to be dictionary encoded unless we're casting the frame to pandas. So can we just flag the columns we want to cast to strings in the future and do so only when conversion to pandas happens?

Comment (AndreyPavlenko) [reply modin/experimental/sql/hdk/query.py:73]: Initially, I've implemented the cast only in to_pandas(), but the SQL tests failed. The problem here is that we don't know the real types.

Comment (AndreyPavlenko) [reply modin/experimental/sql/hdk/query.py:73]: I measured the cast time on my laptop, running WSL2, using the below code:
```python
from modin.experimental.sql import query
data = {str(k): [str(v) for v in range(10000)] for k in range(100)}
df = pd.DataFrame(data)
print(query("SELECT * FROM df", df=df))
```
Time: 0.01735997200012207

Comment (AndreyPavlenko) [reply modin/experimental/sql/hdk/query.py]: Replaced cast with fake dtypes.

**Comment by ienkovich** [comment]
The `to_pandas` part looks good, but I'm really not sure we should try to make a similar thing for SQL queries. Trying to match column names in the original frame and the result to borrow `dtype` might produce really unexpected results. If SQL is used to deliberately produce a column with a new type keeping the old column name, then the introduced code would change the type back to the original.

**Comment by AndreyPavlenko** [comment]
> If SQL is used to deliberately produce a column with a new type keeping the old column name, then the introduced code would change the type back to the original.

No, it shouldn't changed back to the original. This code changes dtypes from category to object for the dictionary fields only. Additionally, it checks - if the field exists in the original frame and the type is categorical, then dtype is not changed for this field. An unexpected result could be if a categorical field is returned with a different name. In this case, it will be casted to string.

**Comment by ienkovich** [comment]
> No, it shouldn't changed back to the original. This code changes dtypes from category to object for the dictionary fields only. Additionally, it checks - if the field exists in the original frame and the type is categorical, then dtype is not changed for this field. An unexpected result could be if a categorical field is returned with a different name. In this case, it will be casted to string.

That's good and reduces the possibility of an undesired cast. But still, it's so easy to get not converted dict simply because names don't match (and any operation on a column modifies its name). Also, we can always come up with a case where category becomes a string with no obvious reason. E.g. columns renaming with a filter:
```
df = pd.DataFrame({"a": ["str1", None, "str2"], "b": ["str3, "str4", None]}
df["b"] = df["b"].astype("category")
query("SELECT b as a, a as b FROM t WHERE b IS NOT NULL;", t=df)
```
I know it's kind of an artificial example, but the same as "SELECT *".

So, does it worth trying to set up dtypes just to cover some cases, break some other cases, and do nothing in probably most of the real scenarios?

**Comment by AndreyPavlenko** [comment]
Yes, I understand it's quite easy to get an unexpected dtype. This is mentioned in the comments. But what is the best solution? I see at least 3:

1. Do nothing here. In this case all string columns are converted to categorical.
2. Convert all dictionaries to string. In this case we miss all categorical columns.
3. Convert all dictionaries to string, but try to keep categoricals.

In my opinion, the first solution is just broken. Here is an example:
```python
df = pd.DataFrame({"a": ["a"]})
res = query("SELECT a FROM t;", t=df) + "_suffix"
```
```
TypeError: unsupported operand type(s) for +: 'Categorical' and 'str'
```

The second one brokes all categorical dtypes.
The last one should work properly in most? real cases, where the column names are not changed, otherwise, we are getting the second scenario.

I think, we should choose between 2 and 3.

**Comment by ienkovich** [comment]
I would choose option 2 for now because it is simple and predictable. Then maybe switch to option 1 when we have reasonable string support in HDK.

**Comment by AndreyPavlenko** [comment]
> I would choose option 2 for now because it is simple and predictable. Then maybe switch to option 1 when we have reasonable string support in HDK.

Thanks, implemented the option 2.

**Review by ienkovich — APPROVED** [review]

File: modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py [hunk 2]
```diff
@@ -2288,7 +2294,25 @@ def to_pandas(self):
...
+            # If the table is exported from HDK, the string columns are converted
+            # to dictionary. On conversion to pandas, these columns will be of type
+            # Categorical, that is not correct. To make the valid conversion, these
+            # fields are cast to string.
+            at = self._partitions[0][0].get()
+            schema = at.schema
+            cast = {
+                i: s[0].name
+                for i, s in enumerate(zip(schema, self._dtypes))
+                if is_dictionary(s[0].type) and not is_categorical_dtype(s[1])
```
Comment (dchigarev) [inline modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py:2307]:
```suggestion
                idx: arrow_type.name
                for idx, (arrow_type, pandas_type) in enumerate(zip(schema, self._dtypes))
                if is_dictionary(arrow_type.type) and not is_categorical_dtype(pandas_type)
```

File: modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py [hunk 3]
```diff
@@ -2288,7 +2294,25 @@ def to_pandas(self):
...
+            }
+            if cast:
+                for i, n in cast.items():
+                    schema = schema.set(i, pyarrow.field(n, pyarrow.string()))
```
Comment (dchigarev) [inline modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py:2311]:
```suggestion
                for idx, new_type in cast.items():
                    schema = schema.set(idx, pyarrow.field(new_type, pyarrow.string()))
```

File: modin/experimental/sql/hdk/query.py [hunk 4]
```diff
@@ -66,7 +69,17 @@ def hdk_query(query: str, **kwargs) -> pd.DataFrame:
     worker = HdkWorker()
     if len(kwargs) > 0:
         query = _build_query(query, kwargs, worker.import_arrow_table)
-    return from_arrow(worker.executeDML(query))
+    df = from_arrow(worker.executeDML(query))
+    mdf = df._query_compiler._modin_frame
+    schema = mdf._partitions[0][0].get().schema
+    # HDK returns strings as dictionary. For the proper conversion to
+    # Pandas, we need to replace dtypes of the corresponding columns.
+    if replace := [i for i, f in enumerate(schema) if pa.types.is_dictionary(f.type)]:
```
Comment (dchigarev) [inline modin/experimental/sql/hdk/query.py:77]:
```suggestion
    if replace := [i for i, col in enumerate(schema) if pa.types.is_dictionary(col.type)]:
```

**Review by dchigarev — APPROVED** [review]
LGTM!

Left a few comments regarding naming.

There are also several category-hacks that still exist in HDK tests, could you please raise an issue to remove the hacks in a separate PR?
https://github.com/modin-project/modin/blob/99f9ed17e76707ecc91c2a69be4a79c35be5ede9/modin/experimental/core/execution/native/implementations/hdk_on_native/test/test_dataframe.py#L1351
https://github.com/modin-project/modin/blob/99f9ed17e76707ecc91c2a69be4a79c35be5ede9/modin/experimental/core/execution/native/implementations/hdk_on_native/test/test_dataframe.py#L2133-L2137
https://github.com/modin-project/modin/blob/99f9ed17e76707ecc91c2a69be4a79c35be5ede9/modin/experimental/core/execution/native/implementations/hdk_on_native/test/test_dataframe.py#L1015
https://github.com/modin-project/modin/blob/99f9ed17e76707ecc91c2a69be4a79c35be5ede9/modin/experimental/core/execution/native/implementations/hdk_on_native/test/test_dataframe.py#L1033-L1034
https://github.com/modin-project/modin/blob/99f9ed17e76707ecc91c2a69be4a79c35be5ede9/modin/experimental/core/execution/native/implementations/hdk_on_native/test/test_dataframe.py#L1134-L1136

**Comment by AndreyPavlenko** [comment]
> There are also several category-hacks that still exist in HDK tests, could you please raise an issue to remove the hacks in a separate PR?

https://github.com/modin-project/modin/issues/5818

**Review by dchigarev — APPROVED** [review]

**Pull request merged**